docs(recipes): update internationalization (#77)
americanexpress/one-app: docs/recipes/Internationalizing-Your-Module.md

# Internationalizing Your Module

Internationalization with [react-intl](https://github.com/formatjs/react-intl) is quick and simple in One App.

> *If you used the [One App Module Generator](https://github.com/americanexpress/one-app-cli/tree/master/packages/generator-one-app-module) to create the module some of these steps may already be completed.*

Reviewer: can we bold this somehow, I cannot imagine anyone not using the generator and this tutorial adds a lot of boilerplate steps, which is good for explanation but not if I miss this note.

Reviewer: and maybe an additional, we highly recommend using the generator and using these steps to explain each part

Author: @jsmartfo Updated, take a look.

## Component Creation
Begin with a simple parent and child component.

`Parent.jsx`
```es6
import React from 'react';
import Child from './Child';

const Parent = () => (
  <Child />
);

export default Parent;
```

`Child.jsx`
```es6
import React from 'React';

const Child = () => <div>greeting</div>;

export default Child;
```

## Module Locale Files

Next create the locale files. [one-app-locale-bundler](https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-locale-bundler) will bundle the locale files as a part of [one-app-bundler](https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-bundler). The README provides more information on the structure of locale files.

> Create a folder structure and content as follows:

Reviewer: >Create a folder structure and content as follows:

>Create a folder structure and content as follows (locale must follow BCP-47):

maybe a small note around locale naming

Author: Done

```
- locale
  - en-US.json
  - es-MX.json
```

`en-US.json`
```json
{
  "greeting": "Welcome!"
}
```

`es-MX.json`
```json
{
  "greeting": "Bienvenidos!"
}

> ```

Reviewer: NOTE: section maybe needed. 

@10xLaCroixDrinker is this still applicable?

The labels `isLoading`, `lastError`, `error`, `locale`, and `links` are reserved and may not be used for application data.

I thought we had a fix for that ...

Author: @jsmartfo I'm confused. The only value for the locale files is the key `greeting`.

Reviewer 2: I believe that is fixed in v4.0.0 of one-app-ducks. the shape of state was changed so that locale data does not mix with the loading data/

Reviewer 3: Yes that was fixed

Reviewer 4: Do we want @brantmullinix to mention that in this documentation?

Author: If it is fixed, not sure we need to mention it. Might just confuse people.

## Module Creation

Turn the `Parent` component into a module. The parent module will use [compose](https://redux.js.org/api/compose/) from [Redux](https://redux.js.org/) and `holocronModule` from [holocron](https://github.com/americanexpress/holocron/tree/master/packages/holocron).

Install Redux and Holocron:
```
  npm i -s redux holocron
```

> Wrap the `Parent` component with the HOC functions. Replace `module-name` with the name of the module.

Reviewer: >HOC

>Higher-Order Component (HOC)

add the definition as first usage

Author: Done

# Internationalizing Your Module

> Internationalization with [react-intl](https://github.com/formatjs/react-intl) is quick and simple in One App. **In fact, if you created your module with the [One App Module Generator](https://github.com/americanexpress/one-app-cli/tree/master/packages/generator-one-app-module) (highly recommended), most of the work is already done!** You can skip ahead to ahead to [Accessing Intl in Child Components](#-accessing-intl-in-child-components). 

Reviewer 5: ```suggestion
This guide demonstrates how to get started with Internationalization with [react-intl](https://github.com/formatjs/react-intl). **In fact, if you created your module with the [One App Module Generator](https://github.com/americanexpress/one-app-cli/tree/master/packages/generator-one-app-module) (highly recommended), most of the work is already done!** You can skip ahead to ahead to [Accessing Intl in Child Components](#-accessing-intl-in-child-components). 
```

However, if you are creating a module from scratch or to better understand the pieces of internationalization in a module, begin at [Component Creation](#-component-creation).

## Component Creation

> Begin with a simple parent and child component.

Reviewer 5: ```suggestion
Begin with a parent and child component.
```

Install One App Ducks:
```
npm i -s @americanexpress/one-app-ducks
```

Load the language pack during load:
```es6
import React from 'react';
import { compose } from 'redux';
import { loadLanguagePack } from '@americanexpress/one-app-ducks';
import Child from './Child';

const Parent = () => (
  <Child />
);

export const loadModuleData = ({ store: { dispatch } }) => dispatch(loadLanguagePack('module-name', { fallbackLocale: 'en-US' }));

Parent.holocron = {
  name: 'module-name',
  loadModuleData
};

export default Parent;
```

## Access Locale and Language Pack
Utilizing [mapStateToProps](https://react-redux.js.org/api/connect#mapstatetoprops-state-ownprops-object) will allow accessing the langugage pack and locale from the Redux state and provide them as props in the `Parent` component.

Install React-Redux, Immutable, and PropTypes:
```
npm i -s redux react-redux immutable prop-types
```

Add a `mapStateToProps` functionality and update the `Parent` component's props.
```es6
import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';

> import { holocronModule } from 'holocron';

Reviewer 6: ```suggestion
```
I believe this can be removed

# Internationalizing Your Module

## Module Locale Files

[one-app-locale-bundler](https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-locale-bundler) will bundle the locale files as a part of [one-app-bundler](https://github.com/americanexpress/one-app-cli/tree/master/packages/one-app-bundler). The [README](https://github.com/americanexpress/one-app-cli/blob/master/packages/one-app-locale-bundler/README.md) provides more information on the structure of locale files.

## Utilizing Language Packs

> Each module must initialize it's own `IntlProvider`. [react-intl](https://github.com/formatjs/react-intl) version 3.x.x is compatible with One App and can be used to initialize an `IntlProvider` with a locale and language pack.

Reviewer 6: With guided tutorials, we need to be more step by step. Any chance we can show the `npm install react-intl@3.x` command and what folder to do it in.

Reviewer 4: @mtomcal, so we don't support `react-intl@4`? Is there a reason for that?

Reviewer 7: It didnt exist when we released one app v5...

> To load a language pack leverage [loadLanguagePack](https://github.com/americanexpress/one-app-ducks) when initializing your module with [holocronModule](https://github.com/americanexpress/holocron).

Reviewer 6: That's great you cited the method libraries. Could we also tighten it up to header links? (e.g. https://github.com/americanexpress/holocron/blob/master/packages/holocron/API.md#holocronmodule)

Author: Sure

With the language pack loaded you can access the state for both the active locale and language pack and use those to create an `IntlProvider`.

```jsx
const Parent = ({ languageData, localeName }) => (
  <IntlProvider locale={localeName} messages={languageData}>
    <Child />
  </IntlProvider>
);

const mapStateToProps = (state) => {
  const localeName = state.getIn(['intl', 'activeLocale']);
  const languagePack = state.getIn(
    ['intl', 'languagePacks', localeName, 'module-name'],
    fromJS({})
  ).toJS();

  return {
    languageData: languagePack && languagePack.data ? languagePack.data : {},
    localeName,
  };
};

export const load = () => (dispatch) => dispatch(loadLanguagePack('module-name', { fallbackLocale: 'en-US' }));

export default compose(
  connect(mapStateToProps),
  holocronModule({
    name: 'module-name',
    load,
  })

> )(Parent);

Reviewer 6: Just FYI we will be introducing an API change for `holocronModule` to become `MyModuleComponent.holocron = { name, load };`. See this PR for some insights: https://github.com/americanexpress/holocron/pull/21

Author: @mtomcal Yup, updating this now.

Author: @mtomcal Updated in last commit.

```

With the `IntlProvider` created at the top level of the module, accessing the language pack is much simpler in child components.

### Higer Order Component

```jsx

> const Child = injectIntl(({ intl }) => <Fragment>{intl.messages.greeting}</Fragment>);

Reviewer 6: Need to show what to `import { injectIntl } from 'react-intl'` in this guided step-by-step tutorial. Could you update all the examples how these methods are imported?

Author: Done
